In [1]:
import pandas as pd
import os

In [14]:
df_ens_preds = pd.read_csv(r"\\zkh\appdata\RTDicom\Projectline_HNC_modelling\Users\Daniel MacRae\1. MultiTox_HNC\Feb_2025_results\Trial_32\ensemble_predictions.csv", sep=';')



df_dataset = pd.read_csv(r"\\zkh\appdata\RTDicom\Projectline_HNC_modelling\Users\Daniel MacRae\1. MultiTox_HNC\DL_NTCP_Multitox\datasets\MT_dataset\stratified_sampling_test_542.csv", sep=';')

df_merged = df_ens_preds.merge(df_dataset, how='left', left_on='PatientID', right_on='PatientID')
df_merged

,PatientID,Mode,Aspiration_M06_pred,Aspiration_M06_true,Dysphagia_M06_pred,Dysphagia_M06_true,Sticky_M06_pred,Sticky_M06_true,Taste_M06_pred,Taste_M06_true,...,Taste_W01_Helemaal_niet,Taste_W01_Nogal,Taste_W01_Nogal_Heel_erg,Dysphagia_W01_Grade0_1,Dysphagia_W01_Grade2,Dysphagia_W01_Grade3_4,Dysphagia_W01_Nogal,Dysphagia_W01_Heel_erg,Dysphagia_W01_Een_beetje,Dysphagia_W01_Helemaal_niet
0,52277,test,0.043978,0.0,0.400356,-1.0,0.289286,0.0,0.263543,1.0,...,1,0,0,0,1,0,0,0,0,0
1,66593,test,0.160205,0.0,0.231431,0.0,0.363443,0.0,0.277286,0.0,...,1,0,0,1,0,0,0,0,0,0
2,92560,test,0.044011,0.0,0.227443,0.0,0.287770,0.0,0.259986,0.0,...,1,0,0,1,0,0,0,0,0,0
3,163517,test,0.027100,0.0,0.033533,0.0,0.109515,0.0,0.052217,0.0,...,1,0,0,1,0,0,0,0,0,0
4,213240,test,0.065761,0.0,0.368783,0.0,0.360731,1.0,0.368292,0.0,...,1,0,0,1,0,0,0,0,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
211,9586209,test,0.057228,0.0,0.539665,1.0,0.341252,0.0,0.314760,0.0,...,0,0,0,0,1,0,0,0,0,0
212,9597467,test,0.061236,0.0,0.258840,1.0,0.321724,0.0,0.307668,1.0,...,1,0,0,1,0,0,0,0,0,0
213,9653665,test,0.099694,0.0,0.254165,0.0,0.355672,1.0,0.338638,1.0,...,1,0,0,1,0,0,0,0,0,0
214,9715913,test,0.090030,1.0,0.449579,1.0,0.422078,1.0,0.425273,1.0,...,1,0,0,1,0,0,0,0,0,0


In [13]:
df_new_ens_preds = pd.read_csv(r'\\zkh\appdata\RTDicom\Projectline_HNC_modelling\Users\Daniel MacRae\1. MultiTox_HNC\Feb_2025_new_patients\ensemble_predictions_PRIMA.csv', sep=';')

df_new_dataset = pd.read_csv(r"\\zkh\appdata\RTDicom\Projectline_HNC_modelling\Users\Daniel MacRae\1. MultiTox_HNC\Dataset\MT_post_2021_validation_set.csv", sep=';')

df_new_merged = df_new_ens_preds.merge(df_new_dataset, how='left', left_on='PatientID', right_on='PatientID')
df_new_merged

,PatientID,Mode,Aspiration_M06_pred,Aspiration_M06_true,Dysphagia_M06_pred,Dysphagia_M06_true,Sticky_M06_pred,Sticky_M06_true,Taste_M06_pred,Taste_M06_true,...,N1,N2,N3,"Roken_Ja, in verleden","Roken_Ja, nog steeds",Roken_Nee,Chemoradiation,Bioradiation,Concurrent_Chemo_or_Bioradiation,Protons
0,10785,test,0.023329,0.0,0.044616,0.0,0.136502,0.0,0.077665,0.0,...,0,0,0,1,0,0,0,0,0,0
1,11247,test,0.049947,0.0,0.148561,1.0,0.260811,0.0,0.224639,1.0,...,0,1,0,0,1,0,0,0,0,0
2,11349,test,0.022382,0.0,0.036705,0.0,0.120605,1.0,0.058891,1.0,...,0,0,0,0,1,0,0,0,0,0
3,11391,test,0.076913,0.0,0.165567,0.0,0.369344,1.0,0.281757,0.0,...,0,1,0,1,0,0,0,0,0,0
4,11481,test,0.087692,0.0,0.188745,0.0,0.372712,1.0,0.316459,0.0,...,0,0,0,1,0,0,0,0,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
270,99577,test,0.025787,0.0,0.094960,0.0,0.177146,0.0,0.114633,0.0,...,0,1,0,0,0,1,0,0,0,1
271,99636,test,0.045301,0.0,0.046652,0.0,0.195166,1.0,0.093605,0.0,...,1,0,0,0,0,1,1,0,1,1
272,99777,test,0.109630,1.0,0.158437,0.0,0.369763,1.0,0.307408,1.0,...,0,1,0,1,0,0,0,0,0,1
273,99870,test,0.085751,-1.0,0.121471,0.0,0.329495,-1.0,0.251048,-1.0,...,0,1,0,1,0,0,0,0,0,1


,PatientID,Mode,Aspiration_M06_pred,Aspiration_M06_true,Dysphagia_M06_pred,Dysphagia_M06_true,Sticky_M06_pred,Sticky_M06_true,Taste_M06_pred,Taste_M06_true,...,N1,N2,N3,"Roken_Ja, in verleden","Roken_Ja, nog steeds",Roken_Nee,Chemoradiation,Bioradiation,Concurrent_Chemo_or_Bioradiation,Protons
0,10785,test,0.023329,0.0,0.044616,0.0,0.136502,0.0,0.077665,0.0,...,0,0,0,1,0,0,0,0,0,0
1,11247,test,0.049947,0.0,0.148561,1.0,0.260811,0.0,0.224639,1.0,...,0,1,0,0,1,0,0,0,0,0
2,11349,test,0.022382,0.0,0.036705,0.0,0.120605,1.0,0.058891,1.0,...,0,0,0,0,1,0,0,0,0,0
3,11391,test,0.076913,0.0,0.165567,0.0,0.369344,1.0,0.281757,0.0,...,0,1,0,1,0,0,0,0,0,0
4,11481,test,0.087692,0.0,0.188745,0.0,0.372712,1.0,0.316459,0.0,...,0,0,0,1,0,0,0,0,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
270,99577,test,0.025787,0.0,0.094960,0.0,0.177146,0.0,0.114633,0.0,...,0,1,0,0,0,1,0,0,0,1
271,99636,test,0.045301,0.0,0.046652,0.0,0.195166,1.0,0.093605,0.0,...,1,0,0,0,0,1,1,0,1,1
272,99777,test,0.109630,1.0,0.158437,0.0,0.369763,1.0,0.307408,1.0,...,0,1,0,1,0,0,0,0,0,1
273,99870,test,0.085751,-1.0,0.121471,0.0,0.329495,-1.0,0.251048,-1.0,...,0,1,0,1,0,0,0,0,0,1


In [9]:
from sklearn.metrics import roc_auc_score


# compute the AUCs of subgroups
def remove_missing_labels(df, label_col):
    """
    remove rows with missing labels from the dataframe. missing labels are equal to -1
    
    """

    df_temp = df.copy()
    df_temp = df_temp[df_temp[label_col] != -1]

    return df_temp

def compute_auc(df, label_col, pred_col):
    """
    compute the AUC of a given label and prediction column in the dataframe
    """
    # remove rows with missing labels
    df_temp = remove_missing_labels(df, label_col)

    # compute the AUC
    auc = roc_auc_score(df_temp[label_col], df_temp[pred_col])

    return auc

# compare groups

In [18]:
list(df_merged.columns)

['PatientID',
 'Mode',
 'Aspiration_M06_pred',
 'Aspiration_M06_true',
 'Dysphagia_M06_pred',
 'Dysphagia_M06_true',
 'Sticky_M06_pred',
 'Sticky_M06_true',
 'Taste_M06_pred',
 'Taste_M06_true',
 'Xerostomia_M06_pred',
 'Xerostomia_M06_true',
 'Sex',
 'Age',
 'CT+C_available',
 'CT_Artefact',
 'Loctum2',
 'Photons',
 'T_stage',
 'N_stage',
 'Smoking',
 'P16',
 'WHO',
 'Xerostomia_BSL',
 'Xerostomia_W01',
 'Aspiration_BSL',
 'Aspiration_W01',
 'Sticky_BSL',
 'Sticky_W01',
 'Taste_BSL',
 'Taste_W01',
 'Dysphagia_BSL',
 'Dysphagia_W01',
 'Xerostomia_M06',
 'Xerostomia_M12',
 'Xerostomia_M18',
 'Aspiration_M06',
 'Aspiration_M12',
 'Aspiration_M18',
 'Sticky_M06',
 'Sticky_M12',
 'Sticky_M18',
 'Taste_M06',
 'Taste_M12',
 'Taste_M18',
 'Dysphagia_M06',
 'Dysphagia_M12',
 'Dysphagia_M18',
 'Parotid_L_meandose',
 'Parotid_R_meandose',
 'Parotid_meandose',
 'Submandibular_L_meandose',
 'Submandibular_R_meandose',
 'Submandibular_meandose',
 'PCM_Sup_meandose',
 'PCM_Med_meandose',
 'PCM_Inf_m

In [ ]:
Loctum2
Photons
CT_Artefact

Smoking_1
P16_Positief
P16_Negatief


In [ ]:
# compute the AUCs of the dataframes, using the groupby function on the column "Technique"
endpoints = ["Aspiration_M06", 'Dysphagia_M06', 'Sticky_M06', 'Taste_M06', 'Xerostomia_M06']

column_name_group = "Sex"

for key, group in df_merged.groupby(column_name_group):
    result_dict = {}
    # compute the AUC of the group
    for endpoint in endpoints:
        pred_col_name = f"{endpoint}_pred"
        true_col_name = f"{endpoint}_true"

        auc = compute_auc(group, true_col_name, pred_col_name)

        result_dict[endpoint] = round(auc, 2)

    print(column_name_group, key, result_dict)

print("- "* 20)

if column_name_group == "Photons": column_name_group = "Protons"
elif column_name_group == "Smoking_1": column_name_group = "Roken"

for key, group in df_new_merged.groupby(column_name_group):
    result_dict = {}
    # compute the AUC of the group
    for endpoint in endpoints:
        pred_col_name = f"{endpoint}_pred"
        true_col_name = f"{endpoint}_true"

        auc = compute_auc(group, true_col_name, pred_col_name)

        result_dict[endpoint] = round(auc, 2)

    print(column_name_group, key, result_dict)

Sex 0 {'Aspiration_M06': 0.79, 'Dysphagia_M06': 0.81, 'Sticky_M06': 0.64, 'Taste_M06': 0.51, 'Xerostomia_M06': 0.7}
Sex 1 {'Aspiration_M06': 0.7, 'Dysphagia_M06': 0.8, 'Sticky_M06': 0.73, 'Taste_M06': 0.71, 'Xerostomia_M06': 0.8}
- - - - - - - - - - - - - - - - - - - - 
Concurrent_Chemo_or_Bioradiation 0 {'Aspiration_M06': 0.53, 'Dysphagia_M06': 0.74, 'Sticky_M06': 0.68, 'Taste_M06': 0.58, 'Xerostomia_M06': 0.69}
Concurrent_Chemo_or_Bioradiation 1 {'Aspiration_M06': 0.58, 'Dysphagia_M06': 0.95, 'Sticky_M06': 0.49, 'Taste_M06': 0.75, 'Xerostomia_M06': 0.59}


In [35]:
list(df_new_merged.columns)

['PatientID',
 'Mode',
 'Aspiration_M06_pred',
 'Aspiration_M06_true',
 'Dysphagia_M06_pred',
 'Dysphagia_M06_true',
 'Sticky_M06_pred',
 'Sticky_M06_true',
 'Taste_M06_pred',
 'Taste_M06_true',
 'Xerostomia_M06_pred',
 'Xerostomia_M06_true',
 'Split',
 'DB',
 'RTSTART',
 'Sex',
 'Age',
 'Loctum',
 'Loctum2',
 'Technique',
 'Inplancode',
 'Systemic_treatment',
 'Modality',
 'Roken',
 'T_stage',
 'M_stage',
 'N_stage',
 'Histology',
 'P16',
 'Doelgebied',
 'WHO',
 'Alcohol',
 'Sticky_BSL',
 'Sticky_W01',
 'Sticky_W02',
 'Sticky_W03',
 'Sticky_W04',
 'Sticky_W05',
 'Sticky_W06',
 'Sticky_W07',
 'Sticky_W12',
 'Sticky_M06',
 'Sticky_M12',
 'Sticky_M18',
 'Sticky_M24',
 'Sticky_M36',
 'Sticky_M48',
 'Sticky_M60',
 'Xerostomia_BSL',
 'Xerostomia_W01',
 'Xerostomia_W02',
 'Xerostomia_W03',
 'Xerostomia_W04',
 'Xerostomia_W05',
 'Xerostomia_W06',
 'Xerostomia_W07',
 'Xerostomia_W12',
 'Xerostomia_M06',
 'Xerostomia_M12',
 'Xerostomia_M18',
 'Xerostomia_M24',
 'Xerostomia_M36',
 'Xerostomia_M4

In [ ]:
Technique {'Aspiration_M06': 0.63, 'Dysphagia_M06': 0.78, 'Sticky_M06': 0.65, 'Taste_M06': 0.55, 'Xerostomia_M06': 0.63}
Technique {'Aspiration_M06': 0.48, 'Dysphagia_M06': 0.69, 'Sticky_M06': 0.7, 'Taste_M06': 0.62, 'Xerostomia_M06': 0.75}

In [ ]:
Technique {'Aspiration_M06': 0.63, 'Dysphagia_M06': 0.78, 'Sticky_M06': 0.65, 'Taste_M06': 0.55, 'Xerostomia_M06': 0.63}


# old
Photons {'Aspiration_M06': 0.75, 'Dysphagia_M06': 0.85, 'Sticky_M06': 0.72, 'Taste_M06': 0.68, 'Xerostomia_M06': 0.82}
Protons {'Aspiration_M06': 0.61, 'Dysphagia_M06': 0.59, 'Sticky_M06': 0.49, 'Taste_M06': 0.48, 'Xerostomia_M06': 0.47}


# new 
Photons {'Aspiration_M06': 0.48, 'Dysphagia_M06': 0.69, 'Sticky_M06': 0.7, 'Taste_M06': 0.62, 'Xerostomia_M06': 0.75}
Protons {'Aspiration_M06': 0.63, 'Dysphagia_M06': 0.78, 'Sticky_M06': 0.65, 'Taste_M06': 0.55, 'Xerostomia_M06': 0.63}